File: 02/a/main.cpp
#include <cstdint>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>

char parse_opponent_move(const char c) {
  switch (c) {
    case 'A':
      return 'R';
    case 'B':
      return 'P';
    case 'C':
      return 'S';
    default:
      throw "bad input";
  }
}

char parse_my_move(const char c) {
  switch (c) {
    case 'X':
      return 'R';
    case 'Y':
      return 'P';
    case 'Z':
      return 'S';
    default:
      throw "bad input";
  }
}

static const std::unordered_map<char, char> winning{
  {'R', 'S'},
  {'P', 'R'},
  {'S', 'P'}
};

static const std::unordered_map<char, std::size_t> score_table{
  {'R', 1},
  {'P', 2},
  {'S', 3}
};

static constexpr std::size_t WIN{6}, TIE{3}, LOSS{};

int main(const int, const char** argv) {
  std::ifstream input_file_stream{argv[1]};
  std::string row;
  std::size_t score{};

  while (std::getline(input_file_stream, row)) {
    const auto my_move{parse_my_move(row.back())};
    const auto opponent_move{parse_opponent_move(row.front())};

    score += score_table.at(my_move);

    if (my_move == opponent_move) {
      score += TIE;
    } else {
      score += winning.at(my_move) == opponent_move ? WIN : LOSS;
    }
  }
  std::cout << score << '\n';
}

File: 02/a/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

var winning = map[byte]byte{
	'R': 'S',
	'P': 'R',
	'S': 'P',
}

var scoreTable = map[byte]uint64{
	'R': 1,
	'P': 2,
	'S': 3,
}

const (
	WIN  uint64 = 6
	TIE  uint64 = 3
	LOSS uint64 = 0
)

func parseOpponentMove(c byte) (byte, error) {
	switch c {
	case 'A':
		return 'R', nil
	case 'B':
		return 'P', nil
	case 'C':
		return 'S', nil
	}
	return 0, fmt.Errorf("unsupported opponent move %c", c)
}

func parseMyMove(c byte) (byte, error) {
	switch c {
	case 'X':
		return 'R', nil
	case 'Y':
		return 'P', nil
	case 'Z':
		return 'S', nil
	}
	return 0, fmt.Errorf("unsupported own move %c", c)
}

func main() {
	var score uint64

	file, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal("couldn't open file")
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		rawText := scanner.Text()

		myMove, myErr := parseMyMove(rawText[len(rawText)-1])
		opponentMove, opponentErr := parseOpponentMove(rawText[0])

		if myErr != nil || opponentErr != nil {
			log.Fatal("Error parsing move")
		}

		score += scoreTable[myMove]

		if myMove == opponentMove {
			score += TIE
		} else if winning[myMove] == opponentMove {
			score += WIN
		} else {
			score += LOSS
		}
	}

	if scanner.Err() != nil {
		log.Fatal("couldn't read file")
	}

	fmt.Println(score)
}
